### PR TITLE
feat: support non-blocking mode for rate limit

### LINKED ISSUE #52: support when token reach the limit for sec, return error code directly rather than block the connection

### PR DESCRIPTION
**Pull request by yjhmelody**
Closed #52

### CODE REVIEW

**Comment by Revolution1** [comment]
typo: no-blocking => non-blocking

File: src/utils/mod.rs [hunk 1]
```diff
@@ -29,6 +29,10 @@ pub mod errors {
         )
     }
 
+    pub fn rate_limit() -> ErrorObjectOwned {
```
Comment (koushiro) [inline src/utils/mod.rs]:
```suggestion
    pub fn reached_rate_limit() -> ErrorObjectOwned {
```

File: src/extensions/rate_limit/mod.rs [hunk 2]
```diff
@@ -80,28 +90,36 @@ impl RateLimitBuilder {
 
         let mut ip_limiter = None;
         let mut ip_jitter = None;
+        let mut ip_no_blocking = false;
```
Comment (koushiro) [inline src/extensions/rate_limit/mod.rs]:
```suggestion
        let mut ip_non_blocking = false;
```

File: src/extensions/rate_limit/mod.rs [hunk 3]
```diff
@@ -80,28 +90,36 @@ impl RateLimitBuilder {
...
             let burst = NonZeroU32::new(rule.burst).unwrap();
             let quota = build_quota(burst, Duration::from_secs(rule.period_secs));
             ip_limiter = Some(Arc::new(RateLimiter::keyed(quota)));
             ip_jitter = Some(Jitter::up_to(Duration::from_millis(rule.jitter_up_to_millis)));
+            ip_no_blocking = rule.non_blocking;
         }
 
         let mut global_limiter = None;
         let mut global_jitter = None;
+        let mut global_no_blocking = false;
```
Comment (koushiro) [inline src/extensions/rate_limit/mod.rs]:
```suggestion
        let mut global_non_blocking = false;
```

File: src/extensions/rate_limit/mod.rs [hunk 4]
```diff
@@ -52,16 +59,19 @@ pub struct RateLimitBuilder {
...
     global_limiter: Option<Arc<DefaultDirectRateLimiter>>,
+    global_no_blocking: bool,
 }
 
 #[async_trait::async_trait]
 impl Extension for RateLimitBuilder {
     type Config = RateLimitConfig;
 
     async fn from_config(config: &Self::Config, _registry: &ExtensionRegistry) -> Result<Self, anyhow::Error> {
+        println!("{:#?}", config);
```
Comment (koushiro) [inline src/extensions/rate_limit/mod.rs]:
```suggestion
```

File: src/extensions/rate_limit/mod.rs [hunk 5]
```diff
@@ -27,16 +27,23 @@ pub struct RateLimitConfig {
...
     // this is to prevent a thundering herd problem https://en.wikipedia.org/wiki/Thundering_herd_problem
     // e.g. if jitter_up_to_millis is 1000, then additional delay of random(0, 1000) milliseconds will be added
     #[serde(default = "default_jitter_up_to_millis")]
     pub jitter_up_to_millis: u64,
+    /// Return a rate limit jsonrpc error directly if true.
+    #[serde(default = "default_no_blocking")]
+    pub non_blocking: bool,
+}
+
+fn default_no_blocking() -> bool {
```
Comment (koushiro) [inline src/extensions/rate_limit/mod.rs]:
```suggestion
fn default_non_blocking() -> bool {
```

File: src/extensions/rate_limit/mod.rs [hunk 6]
```diff
@@ -27,16 +27,23 @@ pub struct RateLimitConfig {
...
     #[serde(default = "default_period_secs")]
     pub period_secs: u64,
     // jitter_millis is the maximum amount of jitter to add to the rate limit
     // this is to prevent a thundering herd problem https://en.wikipedia.org/wiki/Thundering_herd_problem
     // e.g. if jitter_up_to_millis is 1000, then additional delay of random(0, 1000) milliseconds will be added
     #[serde(default = "default_jitter_up_to_millis")]
     pub jitter_up_to_millis: u64,
+    /// Return a rate limit jsonrpc error directly if true.
+    #[serde(default = "default_no_blocking")]
+    pub non_blocking: bool,
```
Comment (koushiro) [inline src/extensions/rate_limit/mod.rs]: how about using `blocking: bool` as option here and it's false by default, and then devops don't need to modify the option explicitly in chart side

Comment (koushiro) [reply src/extensions/rate_limit/mod.rs]: I can change this PR if you agree with it

Comment (yjhmelody) [reply src/extensions/rate_limit/mod.rs]: Both are ok. But should only to change config part.

**Review by koushiro — APPROVED** [review]

**Pull request merged**